File: Router/server/src/services/rag/media.js
import fs from "node:fs";
import path from "node:path";

const MD_IMAGE_RE = /!\[([^\]]*)\]\(([^)\s]+)(?:\s+"[^"]*")?\)/g;
const HTML_IMG_RE = /<img\b[^>]*\bsrc=["']([^"']+)["'][^>]*>/gi;
const ALT_RE = /\balt=["']([^"']*)["']/i;

export function normalizeSrc(src) {
    let cleaned = decodeURIComponent((src || "").trim());
    cleaned = cleaned.replace(/\\/g, "/");
    while (cleaned.startsWith("./"))
        cleaned = cleaned.slice(2);
    return cleaned;
}

export function imageUrl(src) {
    const clean = normalizeSrc(src);
    if (clean.startsWith("assets/"))
        return `/${clean}`;
    if (clean.startsWith("/assets/"))
        return clean;
    return `/assets/${path.basename(clean)}`;
}

export function imageDiskPath(src, assetsDir) {
    const clean = normalizeSrc(src);
    if (clean.startsWith("assets/"))
        return path.join(assetsDir, clean.slice("assets/".length));
    if (clean.startsWith("/assets/"))
        return path.join(assetsDir, clean.slice("/assets/".length));
    return path.join(assetsDir, path.basename(clean));
}

export function extractImageRefs(markdown, assetsDir) {
    const refs = [];
    const seen = new Set();
    const text = markdown || "";

    for (const match of text.matchAll(MD_IMAGE_RE)) {
        const alt = (match[1] || "").trim();
        const src = normalizeSrc(match[2]);
        if (!src || seen.has(src))
            continue;
        seen.add(src);
        const disk = imageDiskPath(src, assetsDir);
        refs.push({
            alt,
            src,
            url: imageUrl(src),
            exists: fs.existsSync(disk),
            kind: "markdown",
        });
    }

    for (const match of text.matchAll(HTML_IMG_RE)) {
        const tag = match[0];
        const src = normalizeSrc(match[1]);
        if (!src || seen.has(src))
            continue;
        seen.add(src);
        const altMatch = ALT_RE.exec(tag);
        const disk = imageDiskPath(src, assetsDir);
        refs.push({
            alt: altMatch ? altMatch[1].trim() : "",
            src,
            url: imageUrl(src),
            exists: fs.existsSync(disk),
            kind: "html",
        });
    }

    return refs;
}

export function refsToDicts(refs) {
    return refs.map((ref) => ({
        alt: ref.alt,
        src: ref.src,
        url: ref.url,
        exists: ref.exists,
        kind: ref.kind,
    }));
}

File: faq/app/media.py
from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path
from urllib.parse import unquote

from .config import Settings, settings


_MD_IMAGE_RE = re.compile(r"!\[([^\]]*)\]\(([^)\s]+)(?:\s+\"[^\"]*\")?\)")
_HTML_IMG_RE = re.compile(r"<img\b[^>]*\bsrc=[\"']([^\"']+)[\"'][^>]*>", re.IGNORECASE)
_ALT_RE = re.compile(r"\balt=[\"']([^\"']*)[\"']", re.IGNORECASE)


@dataclass(frozen=True)
class ImageRef:
    alt: str
    src: str
    url: str
    exists: bool
    kind: str = "image"


def normalize_src(src: str) -> str:
    cleaned = unquote((src or "").strip())
    cleaned = cleaned.replace("\\", "/")
    while cleaned.startswith("./"):
        cleaned = cleaned[2:]
    return cleaned


def image_url(src: str) -> str:
    clean = normalize_src(src)
    if clean.startswith("assets/"):
        return "/" + clean
    if clean.startswith("/assets/"):
        return clean
    return "/assets/" + Path(clean).name


def image_disk_path(src: str, cfg: Settings = settings) -> Path:
    clean = normalize_src(src)
    if clean.startswith("assets/"):
        return cfg.assets_dir / clean[len("assets/") :]
    if clean.startswith("/assets/"):
        return cfg.assets_dir / clean[len("/assets/") :]
    return cfg.assets_dir / Path(clean).name


def extract_image_refs(markdown: str, cfg: Settings = settings) -> list[ImageRef]:
    refs: list[ImageRef] = []
    seen: set[str] = set()

    for match in _MD_IMAGE_RE.finditer(markdown or ""):
        alt = (match.group(1) or "").strip()
        src = normalize_src(match.group(2))
        if not src or src in seen:
            continue
        seen.add(src)
        refs.append(
            ImageRef(
                alt=alt,
                src=src,
                url=image_url(src),
                exists=image_disk_path(src, cfg).is_file(),
                kind="markdown",
            )
        )

    for match in _HTML_IMG_RE.finditer(markdown or ""):
        tag = match.group(0)
        src = normalize_src(match.group(1))
        if not src or src in seen:
            continue
        seen.add(src)
        alt_match = _ALT_RE.search(tag)
        refs.append(
            ImageRef(
                alt=(alt_match.group(1).strip() if alt_match else ""),
                src=src,
                url=image_url(src),
                exists=image_disk_path(src, cfg).is_file(),
                kind="html",
            )
        )

    return refs


def refs_to_dicts(refs: list[ImageRef]) -> list[dict]:
    return [
        {
            "alt": ref.alt,
            "src": ref.src,
            "url": ref.url,
            "exists": ref.exists,
            "kind": ref.kind,
        }
        for ref in refs
    ]
